internal/writeaheadlog/wal.go

```diff
@@ -0,0 +1,295 @@
+	*T
+	Entry
+}] struct {
+	// pointer type trickery above
+	lk sync.Mutex
+
+	path string
+
+	logFiles []logContent[T, PT]
+	active   struct {
+		file       *os.File
+		cborWriter *cbg.CborWriter
+		logContent[T, PT]
+	}
+}
+
+type logContent[T any, PT interface {
+	*T
+	Entry
+}] struct {
+	logname  string
+	content  []T
+	maxEpoch uint64
+	complete bool
+}
+
+func Open[T any, PT interface {
+	*T
+	Entry
+}](directory string) (*WriteAheadLog[T, PT], error) {
+	wal := WriteAheadLog[T, PT]{
+		path: directory,
+	}
+
+	err := wal.readInAll()
+	if err != nil {
+		return nil, fmt.Errorf("reading the WAL: %v", err)
+	}
+
+	return &wal, nil
+}
+
+func (wal *WriteAheadLog[T, PT]) All() []T {
+	wal.lk.Lock()
+	defer wal.lk.Unlock()
+
+	var res []T
+	for _, c := range wal.logFiles {
+		res = append(res, c.content...)
+	}
+	res = append(res, wal.active.content...)
+	return res
+}
+
+/*
+// TODO: version file
+// or mabye not todo, IDK
+const versionFileName = "version"
+const version = 0
+func (wal *WriteAheadLog[T]) readVersion() (int, error) {
+	versionContent, err := os.ReadFile(path.Join(wal.path, versionFileName))
+	if errors.Is(err, os.ErrNotExist) {
+		return -1, nil
+	}
+	if err != nil {
+		return -1, fmt.Errorf("reading the version file: %w", err)
+	}
+
+	version, err := strconv.ParseInt(string(versionContent), 10, 32)
+	if err != nil {
+		return -1, fmt.Errorf("parsing the version number: %w", err)
+	}
+
+	return version, nil
+}
+*/
+
+func (wal *WriteAheadLog[T, PT]) Log(value T) error {
+	wal.lk.Lock()
+	defer wal.lk.Unlock()
+
+	if err := wal.maybeRotate(); err != nil {
+		return fmt.Errorf("attemting to rotate: %w", err)
+	}
+
+	err := PT(&value).MarshalCBOR(wal.active.cborWriter)
+	if err != nil {
+		return fmt.Errorf("saving value to WAL: %w", err)
+	}
+	err = wal.active.file.Sync()
+	if err != nil {
+		return fmt.Errorf("sycning the file: %w", err)
+	}
+
+	wal.active.maxEpoch = max(wal.active.maxEpoch, PT(&value).WALEpoch())
+	wal.active.content = append(wal.active.content, value)
+
+	return nil
+}
+
+// Purge removes files containing entries only order than keepEpoch
+// It is not guarnateed to remove all entries older than keepEpoch as it will keep the current
+// file.
+func (wal *WriteAheadLog[T, PT]) Purge(keepEpoch uint64) error {
+	wal.lk.Lock()
+	defer wal.lk.Unlock()
+
+	var keptLogFiles []logContent[T, PT]
+	var err error
+	for _, c := range wal.logFiles {
+		if c.maxEpoch < keepEpoch {
+			err1 := os.Remove(filepath.Join(wal.path, c.logname))
+			err = multierr.Append(err, fmt.Errorf("removing WAL file %q: %w", c.logname, err1))
+
+			continue
+		}
+		keptLogFiles = append(keptLogFiles, c)
+	}
+	wal.logFiles = keptLogFiles
+	return nil
+}
+
+const roatateAt = 1 << 20 // 1MiB
+
+func (wal *WriteAheadLog[T, PT]) maybeRotate() error {
+	if wal.active.file == nil {
+		return wal.rotate()
+	}
+	stats, err := wal.active.file.Stat()
+	if err != nil {
+		return fmt.Errorf("collecting stats for the file: %w", err)
```

Comment:
IIRC, the "right" way to do this is to seek to the end of the file and look at the return value. We want to be at the end of the file anyways...

Alternatively, we can get our current position by calling `Seek(0, 1)`.